crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -499,6 +519,116 @@ impl<'a> TimelineEventHandler<'a> {
+                    original_sender = ?event_item.sender(), edit_sender = ?self.ctx.sender,
+                    "Edit event applies to another user's timeline item, discarding"
+                );
+                return None;
+            }
+
+            let poll_state = if let TimelineItemContent::Poll(poll_state) = &event_item.content() {
+                poll_state
+            } else {
+                info!(
+                    original_sender = ?event_item.sender(), edit_sender = ?self.ctx.sender,
+                    "Can't edit a poll that is not of type TimelineItemContent::Poll, discarding"
+                );
+                return None;
+            };
+
+            let new_content = if let Ok(edited_poll_state) =
+                poll_state.edit(self.ctx.timestamp, &replacement.new_content)
+            {
+                TimelineItemContent::Poll(edited_poll_state)
+            } else {
+                info!(
+                    original_sender = ?event_item.sender(), edit_sender = ?self.ctx.sender,
+                    "This poll can't be edited, discarding"
+                );
+                return None;
+            };
+
+            let edit_json = match &self.ctx.flow {
+                Flow::Local { .. } => None,
+                Flow::Remote { raw_event, .. } => Some(raw_event.clone()),
+            };
+
+            trace!("Applying edit");
+            Some(event_item.with_content(new_content, edit_json))
+        });
+    }
+
+    fn handle_poll_start(&mut self, c: UnstablePollStartEventContent, should_add: bool) {
+        let mut poll_state = PollState::new(self.ctx.sender.clone(), self.ctx.timestamp, c);
+        if let Flow::Remote { event_id, .. } = self.ctx.flow.clone() {
+            self.state.poll_cache.apply(&event_id, &mut poll_state);
+        };
+        self.add(should_add, TimelineItemContent::Poll(poll_state));
+    }
+
+    fn handle_poll_response(&mut self, c: UnstablePollResponseEventContent) {
+        let found = _update_timeline_item(
+            &mut self.state.items,
+            &mut self.result.items_updated,
+            &c.relates_to.event_id,
+            "poll response",
+            |event_item| match event_item.content() {
+                TimelineItemContent::Poll(poll_state) => Some(event_item.with_content(
+                    TimelineItemContent::Poll(poll_state.add_response(
+                        &self.ctx.sender,
+                        &self.ctx.timestamp,
+                        &c,
+                    )),
+                    None,
+                )),
+                _ => None,
+            },
+        );
+        if !found {
+            self.state.poll_cache.add_response(
+                &c.relates_to.event_id,
+                &self.ctx.sender,
+                &self.ctx.timestamp,
+                &c,
+            );
+        }
+    }
+
+    fn handle_poll_end(&mut self, c: UnstablePollEndEventContent) {
+        let found = _update_timeline_item(
```

Comment:
Please actually copy-paste the body of `_update_timeline_item` instead of calling it here. It makes no sense for this method to log "Timeline item not found, discarding poll end" if the item is not found, since it won't actually be discarded, it will be added to the poll cache. We could put the found branch of `update_timeline_item` in a new function and reuse that though. Maybe *that* should be named `update_timeline_item`, and the other one `find_and_update_timeline_item`.